concourse/pipelines/monitoring.yml

```diff
@@ -0,0 +1,287 @@
+            EOF
+
+            current_task_definition_arn="$(aws ecs describe-services --services "$APPLICATION" --cluster monitoring --region "$AWS_REGION" | jq -r '.services[0].taskDefinition')"
+            if [ -z "${current_task_definition_arn}" ]; then
+              echo "failed to retrieve current task definition for ${APPLICATION}, exiting..."
+              exit 1
+            fi
+
+            new_task_definition_arn="$(cat "terraform-outputs/${APPLICATION}_task_definition_arn")"
+            if [ -z "${new_task_definition_arn}" ]; then
+              echo "failed to retrieve new task definition for ${APPLICATION}, exiting..."
+              exit 1
+            fi
+
+            # This conditional is used to skip `aws ecs update-service` below since its outputs may confuse readers and
+            # lead them to believe that `aws ecs update-service` is not idempotent
+            if [ "${current_task_definition_arn}" = "${new_task_definition_arn}" ]; then
+              echo "No need to update ${APPLICATION} service since its task definition was not updated"
+              exit 0
+            fi
+
+            echo "Updating $APPLICATION service..."
+
+
+            aws ecs update-service \
+              --cluster monitoring \
+              --service "$APPLICATION" \
+              --task-definition "$new_task_definition_arn" \
+              --region "$AWS_REGION"
+
+            echo "Waiting for $APPLICATION ECS service to reach steady state..."
+
+            aws ecs wait services-stable \
+              --cluster govuk \
+              --services "$APPLICATION" \
+              --region "$AWS_REGION"
+
+            echo "Finished updating $APPLICATION to task definition $new_task_definition_arn."
+
+  - name: configure-grafana
+    plan:
+    - get: govuk-infrastructure
+      passed:
+      - deploy-grafana
+      trigger: true
+    - task: terraform-apply
+      config:
+        inputs:
+        - name: govuk-infrastructure
+          path: src
+        outputs:
+        - name: terraform-outputs
+        params:
+          AWS_REGION: eu-west-1
+          ASSUME_ROLE_ARN: 'arn:aws:iam::430354129336:role/govuk-concourse-deployer'
+        platform: linux
+        image_resource:
+          type: docker-image
+          source:
+            repository: governmentpaas/terraform
+            tag: terraform-0.13.3
+            username: ((docker_hub_username))
+            password: ((docker_hub_authtoken))
+        run:
+          path: sh
+          args:
+          - '-c'
+          - |
+            set -eu
+
+            root_dir=$(pwd)
+            echo "${root_dir}"
+
+            INFRA_DIR="${root_dir}/src/terraform/deployments/monitoring-test/infra"
+            cd ${INFRA_DIR}
+            terraform init -backend-config "role_arn=$ASSUME_ROLE_ARN"
+            grafana_fqdn=$(terraform output grafana_fqdn)
+            echo "Grafana base URL: https://${grafana_fqdn}"
+
+            wait_url() {
```

Comment:
It shouldn't be necessary to reimplement readiness checking like this - the point of having healthchecks on the ECS service is to take care of this. It's ok for a quick demo but we'll need to do better for prod. It looks like you already have a healthcheck on Grafana - is it not doing what it needs to? i.e. is `wait services-stable` returning before Grafana is healthy?

Reply:
I think `ecs wait services-stable` should probably work here. It's not just a check for the correct task count.

It waits until the query returns true. The query is [roughly](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/ecs/wait/services-stable.html):

```
length(service.deployments) == 1 && service.runningCount == service.desiredCount
```

Services have a [deployment](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_Deployment.html) list. It looks like each task definition associated with a service is represented by a deployment:

```
            "deployments": [
                {
                    "id": "ecs-svc/3010894344502080939",
                    "status": "PRIMARY",
                    "taskDefinition": "arn:aws:ecs:eu-west-1:430354129336:task-definition/frontend:201",
                    "desiredCount": 1,
                    "pendingCount": 0,
                    "runningCount": 1,
                    ...
```

So the service is stable when:

- the running count of the service matches the desired count, and
- the service has only one associated task definition (i.e. the deployment has completed, the INACTIVE state task definition no longer appears in the list).
 
I'm not sure if this is a guarantee that the new task is registered and healthy with the ALB target group. Perhaps we should open a ticket with AWS support to confirm this, but I think when the old instance is killed (i.e. runningCount == desiredCount) then the new instance should already be registered with the target group, otherwise this wouldn't be a zero downtime deploy for requests coming through the target group. It would be bizarre, but it wouldn't surprise me if it worked this way, it is very poorly documented.

If it is the case that the ALB health check will trigger ECS to kill tasks after the deploy completes, then maybe you could use a combination of `ecs wait services-stable` (deploy finished) and [elbv2 wait target-in-service](https://docs.aws.amazon.com/cli/latest/reference/elbv2/wait/target-in-service.html) (example below) but I think that probably points to us doing something wrong, if we are 1) bringing up a new instance, 2) killing the old one, and then 3) killing the new one.

```
new_task_ips=$(aws ecs describe-tasks ...)
target_group_arn=$(terraform output public_alb_target_group_arn)
aws elbv2 wait target-in-service \
  --target-group-arn $target_group_arn \
  --targets $new_task_ips
```

(Nb on this example, I can imagine this could be quite flaky, it is a good argument for baking the Grafana config into the docker image if we needed to do something like it).